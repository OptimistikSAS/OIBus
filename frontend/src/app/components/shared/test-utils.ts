import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
