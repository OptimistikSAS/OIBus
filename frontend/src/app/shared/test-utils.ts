import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
