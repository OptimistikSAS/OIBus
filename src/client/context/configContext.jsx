// Context for Engine Json
// load engine on startup
// provide then json and a way to update as a reducer
import React from 'react'
import PropTypes from 'prop-types'
import objectPath from 'object-path'
import apis from '../services/apis'
import utils from '../helpers/utils'

const reducer = (state, action) => {
  const { name, value, config, type, validity } = action
  const newState = utils.jsonCopy(state)
  switch (type) {
    case 'reset':
      return utils.jsonCopy(config)
    case 'update':
      if (!newState.errors) newState.errors = {}
      if (validity) {
        newState.errors[name] = validity
      } else {
        // clean the error object
        delete newState.errors[name]
        if (Object.keys(newState.errors).length === 0) delete newState.errors
      }
      objectPath.set(newState, name, value)
      return newState
    case 'deleteRow':
    {
      // clean the error object
      if (newState.errors) {
        const keys = Object.keys(newState.errors).filter((key) => key.startsWith(name))
        keys.forEach((key) => delete newState.errors[key])
        if (Object.keys(newState.errors).length === 0) delete newState.errors
      }
      objectPath.del(newState, name)
      return newState
    }
    case 'deleteAllRows':
      objectPath.empty(newState, name)
      return newState
    case 'addRow':
      objectPath.push(newState, name, value)
      // copy into the new state
      return newState
    case 'importPoints':
      objectPath.set(newState, name, value)
      return newState
    default:
      throw new Error(`unknown action type: ${type}`)
  }
}
const configInitialState = {}
const ConfigContext = React.createContext(configInitialState)
const ConfigProvider = ({ children }) => {
  const [newConfig, dispatchNewConfig] = React.useReducer(reducer, configInitialState)
  const [activeConfig, setActiveConfig] = React.useState(null)
  const [apiList, setApiList] = React.useState()
  const [protocolList, setProtocolList] = React.useState()

  /**
   * Acquire the list of API
   * @returns {void}
   */
  React.useEffect(() => {
    const call = async () => {
      try {
        setApiList(await apis.getNorthApis())
      } catch (error) {
        console.error(error)
      }
    }
    call()
  }, [])

  /**
   * Acquire the list of Protocols
   * @returns {void}
   */
  React.useEffect(() => {
    const call = async () => {
      try {
        setProtocolList(await apis.getSouthProtocols())
      } catch (error) {
        console.error(error)
      }
    }
    call()
  }, [])

  // On mount, acquire the Active config from server.
  React.useEffect(() => {
    let mounted = true
    const fetchActiveConfig = async () => {
      try {
        const resp = await apis.getConfig()
        if (mounted && resp) {
          dispatchNewConfig({ type: 'reset', config: resp.config })
          // parse/stringify to create deep copy:
          setActiveConfig(utils.jsonCopy(resp.config))
        }
      } catch (error) {
        console.error(error)
      }
    }
    fetchActiveConfig()
    return () => {
      mounted = false
    }
  }, [])
  // the provider return the new and active config and their respective setters
  /**
   * @todo: component using this context (.i.e the whole application) will rerender
   * 4 times (when apiList, protocolList, activeConfig and newConfig are updated).
   * we could make a single call to the server to avoid this effect.
   */
  return (
    <ConfigContext.Provider
      value={{ newConfig, dispatchNewConfig, activeConfig, setActiveConfig, apiList, protocolList }}
    >
      {children}
    </ConfigContext.Provider>
  )
}

ConfigProvider.propTypes = { children: PropTypes.element.isRequired }
export { ConfigContext, ConfigProvider, reducer }
