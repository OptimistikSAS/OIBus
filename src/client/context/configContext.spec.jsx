import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import testConfig from '../../../tests/testConfig'
import { ConfigProvider } from './configContext.jsx'

// mock fetch
global.fetch = jest.fn().mockImplementation((uri) => {
  let jsonString
  switch (uri) {
    case '/config/schemas/north':
      jsonString = JSON.stringify(testConfig.apiList)
      break
    case '/config/schemas/south':
      jsonString = JSON.stringify(testConfig.protocolList)
      break
    case '/config':
      jsonString = JSON.stringify({ config: testConfig })
      break
    default:
      jsonString = '""'
  }
  return {
    status: 200,
    text: jest.fn().mockImplementation(() => jsonString),
  }
})

// mock states
let initState = 'activeConfig'
const setActiveConfig = jest.fn()
const setApiList = jest.fn()
const setProtocolList = jest.fn()
const setState = jest.fn()
React.useState = jest.fn().mockImplementation((init) => {
  if (init === null && initState === 'activeConfig') {
    initState = 'apiList'
    return [init, setActiveConfig]
  }
  if (init === undefined && initState === 'apiList') {
    initState = 'protocolList'
    return [init, setApiList]
  }
  if (init === undefined && initState === 'protocolList') {
    initState = 'activeConfig'
    return [init, setProtocolList]
  }
  return [init, setState]
})
// mock reducer
const dispatchNewConfig = jest.fn()
React.useReducer = jest.fn().mockImplementation(() => [testConfig, dispatchNewConfig])

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('ConfigProvider', () => {
  test('check ConfigProvider', () => {
    act(() => {
      ReactDOM.render(
        <ConfigProvider>
          <div />
        </ConfigProvider>,
        container,
      )
    })
    expect(container).toMatchSnapshot()
  })
  test('check initial data setup', async () => {
    await act(async () => {
      ReactDOM.render(
        <ConfigProvider>
          <div />
        </ConfigProvider>,
        container,
      )
    })
    expect(dispatchNewConfig).toBeCalledWith({ type: 'reset', config: testConfig })
    expect(setActiveConfig).toBeCalledWith(testConfig)
    expect(setApiList).toBeCalledWith(testConfig.apiList)
    expect(setProtocolList).toBeCalledWith(testConfig.protocolList)
    expect(container).toMatchSnapshot()
  })
})
