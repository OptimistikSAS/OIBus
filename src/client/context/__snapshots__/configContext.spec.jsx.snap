// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider check ConfigProvider 1`] = `
<div>
  <div />
</div>
`;

exports[`ConfigProvider check initial data setup 1`] = `
<div>
  <div />
</div>
`;
